src/bundle/Routing/ExpressionLanguage/ContentTypeHeaderMatcherFactory.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Bundle\Rest\Routing\ExpressionLanguage;
+
+use Closure;
+use Ibexa\Contracts\Rest\Input\MediaTypeParserInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final readonly class ContentTypeHeaderMatcherFactory
+{
+    public function __construct(
+        private MediaTypeParserInterface $mediaTypeParser
+    ) {
+    }
+
+    public function __invoke(): Closure
+    {
+        return function (Request $request): ?string {
+            $contentTypeHeaderValue = $request->headers->get('Content-Type');
+
+            if ($contentTypeHeaderValue === null) {
+                return null;
+            }
+
+            return $this->mediaTypeParser->parseContentTypeHeader($contentTypeHeaderValue);
+        };
+    }
```

Comment:
I'm not a huge fan of using closures here as it's not very strict, at least on the outside.
I'm a bit surprised this part doesn't look closer to [that doc](https://symfony.com/doc/5.x/components/expression_language.html#using-expression-providers). Or am I looking at the wrong thing?

Reply:
I guess using a function provider would be safer. Relying on an "internal" Symfony tag (`routing.expression_language_function`) to register our functions in the routing expression engine might break at some point (although unlikely), meanwhile function provider tag is [documented](https://symfony.com/doc/current/reference/dic_tags.html#routing-expression-language-provider).

Using an expression provider might be however more difficult. And they will still contain closures inside their implementation. The question is can we implement it in a timely manner :)

Reply:
@alongosz @Steveb-p would it be even possible to register such a provider under the `routing`? 

besides, it's kinda documented in the Changelog - https://github.com/symfony/framework-bundle/blob/7.1/CHANGELOG.md#510

Reply:
Of course.

`Symfony\Component\Routing\Matcher\ExpressionLanguageProvider` is how Symfony registers it's own functions after all.

Reply:
As discussed on Slack simplified it in https://github.com/ibexa/rest/pull/100/commits/49f399758708ef86ea15f271c9abdddc869c24d7